Mathlib/GroupTheory/GroupAction/Primitive.lean

```diff
@@ -0,0 +1,282 @@
+
+/-- An action is preprimitive if it is pretransitive and
+the only blocks are the trivial ones -/
+@[to_additive]
+class IsPreprimitive [SMul G X] extends IsPretransitive G X : Prop where
+/-- An action is preprimitive if it is pretransitive and
+the only blocks are the trivial ones -/
+  isTrivialBlock_of_isBlock : ∀ {B : Set X}, IsBlock G B → IsTrivialBlock B
+
+open IsPreprimitive
+
+/-- An additive action of an additive group is quasipreprimitive if any normal subgroup
+that has no fixed point acts pretransitively -/
+class _root_.AddAction.IsQuasipreprimitive
+    [AddGroup G] [AddAction G X] extends AddAction.IsPretransitive G X : Prop where
+  isPretransitive_of_normal :
+    ∀ {N : AddSubgroup G} (_ : N.Normal), AddAction.fixedPoints N X ≠ ⊤ →
+      AddAction.IsPretransitive N X
+
+/-- A `MulAction` of a group is quasipreprimitive if any normal subgroup
+that has no fixed point acts pretransitively -/
+@[to_additive]
+class IsQuasipreprimitive [Group G] [MulAction G X] extends IsPretransitive G X : Prop where
+  isPretransitive_of_normal :
+    ∀ {N : Subgroup G} [N.Normal], fixedPoints N X ≠ ⊤ → IsPretransitive N X
+
+variable {G X}
+
+@[to_additive]
+theorem IsPreprimitive.subsingleton_or_eq_univ_of_isBlock
+    [SMul G X] [IsPreprimitive G X] {B : Set X} (hB : IsBlock G B) :
+    B.Subsingleton ∨ B = .univ :=
+  isTrivialBlock_of_isBlock hB
+
+@[to_additive (attr := nontriviality)]
+theorem isPreprimitive_of_subsingleton [SMul G X] [Nonempty G] [Subsingleton X] :
+    IsPreprimitive G X where
+  exists_smul_eq (x y) := by
+    use Classical.arbitrary G
+    rw [eq_iff_true_of_subsingleton]
+    trivial
+  isTrivialBlock_of_isBlock B := by
+    left
+    exact Set.subsingleton_of_subsingleton
+
+variable [Group G] [MulAction G X]
+
+open scoped BigOperators Pointwise
+
+/-- If the action is pretransitive, then the trivial blocks condition implies preprimitivity
+(based condition) -/
+@[to_additive
+"If the action is pretransitive, then the trivial blocks condition implies preprimitivity
+(based condition)"]
+theorem isPreprimitive_of_isTrivialBlock_base [IsPretransitive G X] (a : X)
+    (H : ∀ {B : Set X} (_ : a ∈ B) (_ : IsBlock G B), IsTrivialBlock B) :
+    IsPreprimitive G X where
+  isTrivialBlock_of_isBlock {B} hB := by
+    cases Set.eq_empty_or_nonempty B with
+    | inl h => apply Or.intro_left; simp [h, Set.subsingleton_empty]
+    | inr h =>
+      obtain ⟨b, hb⟩ := h
+      obtain ⟨g, hg⟩ := exists_smul_eq G b a
+      rw [← IsTrivialBlock.smul_iff g]
+      apply H _ (hB.translate g)
+      rw [← hg]
+      use b
+
+/-- If the action is not trivial, then the trivial blocks condition implies preprimitivity
+(pretransitivity is automatic) (based condition) -/
+@[to_additive
+  "If the action is not trivial, then the trivial blocks condition implies preprimitivity
+  (pretransitivity is automatic) (based condition)"]
+theorem isPreprimitive_of_isTrivialBlock_of_not_mem_fixedPoints {a : X} (ha : a ∉ fixedPoints G X)
```

Comment:
```suggestion
theorem IsPreprimitive.of_isTrivialBlock_of_not_mem_fixedPoints {a : X} (ha : a ∉ fixedPoints G X)
```

Reply:
Oh sorry, I didn't see you had reverted this. `IsPreprimitive.of_isTrivialBlock_of_not_mem_fixedPoints` is nice because you get to use anonymous dot notation: It lets you write things like `IsPreprimitive G X := .of_isTrivialBlock_of_not_mem_fixedPoints ...`

Reply:
But then, I should do the same for `…_of_subsingleton`, no?

Reply:
Yes, absolutely! And a few other ones too